src/librustc_data_structures/sync.rs

```diff
@@ -788,3 +789,27 @@ impl<T> DerefMut for OneThread<T> {
         &mut self.inner
     }
 }
+
+/// This is essentially an AtomicPtr but is guaranteed to always be valid
+pub struct AtomicRef<T: 'static>(AtomicPtr<T>, PhantomData<&'static T>);
```

Comment:
This doesn't belong in `sync.rs` since it's not conditionally atomic / thread-safe.